Offline VAD+ASR tutorial

fayejf opened the pull request:
# What does this PR do ?

This PR adds functions and a tutorial about how to perform offline VAD before ASR.
Pipeline: audio -> [VAD] -> speech segments -> [ASR] -> segmented ASR output -> stitched ASR output

**Collection**:  ASR

# Changelog 
- Transcribe partial audio (for example speech segments) with manifest.
- Stitch segmented ASR output.
- Construct the manifest that aligns stitched ASR outputs and ground truth text. 
- And a walkthrough tutorial for the whole process. 

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.


## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

titu1994 commented on examples/asr/transcribe_speech.py:
```diff
@@ -151,10 +151,18 @@ def main(cfg: TranscriptionConfig) -> TranscriptionConfig:
         # get filenames from manifest
         filepaths = []
+        if os.stat(cfg.dataset_manifest).st_size == 0:
+            logging.warning(f"The input dataset_manifest {cfg.dataset_manifest} is empty. Exiting!")
```
Logging.error is better here. Also, cool check to test for empty manifest.

titu1994 reviewed (changes requested):
This is not a good change. the .transcribe() method will only work on python string paths. It is not meant for manifest handling and setting up of lengths etc. That's the point of the transcribe_speech.py script - to provide a higher level API over the transcribe method (supporting manifest etc). 

Please revert these changes to transcribe - especially the requirement of num files etc. Such things are not meant for transcribe. They are meant for higher level script to work on transcribe.

titu1994 commented on examples/asr/transcribe_speech.py:
```diff
@@ -93,6 +94,86 @@ class TranscriptionConfig:
     rnnt_decoding: RNNTDecodingConfig = RNNTDecodingConfig(fused_batch_size=-1)
 
 
+def transcribe_partial_audio(
```
Where's the equivalent for RNNT ? If it's not supported, raise an error please. You can do this by asserting that obj is instance of only CTC classes.

titu1994 commented on examples/asr/transcribe_speech.py:
```diff
@@ -151,10 +232,22 @@ def main(cfg: TranscriptionConfig) -> TranscriptionConfig:
+            partial_audio = True
             for line in f:
                 item = json.loads(line)
+                if "duration" not in item:
```
duration is not the key factor determining limit on audio - it is both duration + offset presence that determines that partial audio is chosen. 

Duration is filled in for nearly all manifests always.

titu1994 commented on examples/asr/transcribe_speech.py:
```diff
@@ -151,10 +232,22 @@ def main(cfg: TranscriptionConfig) -> TranscriptionConfig:
+                if "duration" not in item:
+                    partial_audio = False
                 filepaths.append(item['audio_filepath'])
+        if not partial_audio:
```
Do not raise this warning - by default most manifest do not have offset set. We don't want to warm for a general case

titu1994 commented on examples/asr/transcribe_speech.py:
```diff
@@ -93,6 +94,86 @@ class TranscriptionConfig:
     rnnt_decoding: RNNTDecodingConfig = RNNTDecodingConfig(fused_batch_size=-1)
 
 
+def transcribe_partial_audio(
```
These functions are quite large, better to move the into some util dir and import only here.

fayejf replied:
Cannot find a good existing place. How about having a transcribe_utils.py so we can put everything about transcribe there.

titu1994 commented on nemo/collections/asr/models/ctc_models.py:
```diff
@@ -691,9 +691,15 @@ def _setup_transcribe_dataloader(self, config: Dict) -> 'torch.utils.data.DataLo
             A pytorch DataLoader for the given audio file(s).
         """
-        batch_size = min(config['batch_size'], len(config['paths2audio_files']))
+        if 'manifest_filepath' in config:
```
What about RNNT overrides ? Add the support for it's two methods, even if it's not used yet. All transcribe related info should ideally be synchronized among all model implementations.

fayejf replied:
good point. I totally forgot rnnt. :(

fayejf replied:
added support for rnnt and rnnt bpe

titu1994 replied:
Cool.

titu1994 commented:
The notebook needs some grammar fixes + @tango4j  for review of the VAD code (it looks correct to me but second checks). 

Also, the method does not work for RNNT - the notebook needs to be renamed to ASR with VAD for CTC models (or some such equivalent title) + mention in the notebook that this is only for CTC at the moment. 
Also add this notebook link to the ASR tutorials readme https://github.com/NVIDIA/NeMo/tree/main/tutorials/asr under it's own section (ASR with VAD).

fayejf commented:
> Also, the method does not work for RNNT - the notebook needs to be renamed to ASR with VAD for CTC models (or some such equivalent title) + mention in the notebook that this is only for CTC at the moment. Also add this notebook link to the ASR tutorials readme https://github.com/NVIDIA/NeMo/tree/main/tutorials/asr under it's own section (ASR with VAD).

I added it to readme. [11] Is there any particular order for notebooks?
Oh, good point. it's should be Offline_ASR_with_VAD_for_CTC_models for now.

fayejf commented on nemo/collections/asr/parts/utils/vad_utils.py:
```diff
@@ -881,3 +881,121 @@ def init_vad_model(model_path):
+        stitched_output_manifest = "asr_stitched_output_manifest.json"
+
+    if not speech_segments_tensor_dir:
+        speech_segments_tensor_dir = "speech_segments"
```
good point!

fayejf commented:
@titu1994 @tango4j  Thanks for the careful review and feedback! I've addressed all the feedback above and did a grammar check again. Please review again when you have time. :)

titu1994 commented on docs/source/starthere/tutorials.rst:
```diff
@@ -88,6 +88,9 @@ To run a tutorial:
      - `Buffered Transducer inference with LCS Merge <https://colab.research.google.com/github/NVIDIA/NeMo/blob/stable/tutorials/asr/Buffered_Transducer_Inference_with_LCS_Merge.ipynb>`_
+   * - ASR
+     - Offline ASR with VAD for CTC models
+     - `Offline ASR with VAD for CTC models <https://github.com/NVIDIA/NeMo/blob/stable/tutorials/asr/Offline_ASR_with_VAD_for_CTC_models.ipynb>`_
```
Shouldn't this be a colab link ?

fayejf replied:
good catch!

titu1994 commented on examples/asr/transcribe_speech.py:
```diff
@@ -151,10 +152,18 @@ def main(cfg: TranscriptionConfig) -> TranscriptionConfig:
+            partial_audio = True
             for line in f:
                 item = json.loads(line)
+                if "offset" not in item or "duration" not in item:
```
This should be an if x in manifesf and y in manifest: situation, not an or. Most manifests will have duration but not offset.

fayejf replied:
Not really. if anyone is missing, then partial_audio=False and we will use `audio_filepath` and `transcribe` method

titu1994 replied:
We generally have the 3 values in the manifest - audio_filepath. text and duration. Here the or for "offset" will fail and set partial_audio = False which is fine.
The logic should be - partial audio is false by default, but if both args are provided, then set to true - its the inverse of this.

titu1994 commented on nemo/collections/asr/parts/utils/transcribe_utils.py:
```diff
@@ -0,0 +1,104 @@
+) -> List[str]:
+
+    assert (
+        type(asr_model) == EncDecCTCModelBPE or type(asr_model) == EncDecCTCModel
```
You can do is instance(asr_model, EncDecCTCModel) - subclassing should work out of the box.

titu1994 commented on examples/asr/transcribe_speech.py:
```diff
@@ -187,15 +196,25 @@ def autocast():
+                    batch_size=cfg.batch_size,
+                    num_workers=cfg.num_workers,
+                )
+            else:
```
Hmm we don't want to disallow users who have offset and duration in manifest from running transcribe (at least whole transcribe can work).  

Do one thing. Check for CTC here then if partial audio send to transcribe partial. If not CTC and still partial audio flag is set - log a warning message saying that RNNT models do not support partial audio transcription (for now) and instead perform the else part.

titu1994 commented on tutorials/asr/README.md:
```diff
@@ -28,6 +28,8 @@ In this repository, you will find several tutorials discussing what is Automatic
 
 10) `Self_Supervised_Pre_Training`: It can often be difficult to obtain labeled data for ASR training. In this tutorial, we demonstrate how to pre-train a speech model in an unsupervised manner, and then fine-tune with CTC loss.
 
+11) `Offline_ASR_with_VAD_for_CTC_models`: In this tutorial, we will demonstrate how to use offline VAD to extract speech segments and transcribe the speech segments with CTC models. This will help to exclude some non_speech utterances and could save computation resources by removing unnecessary input to the ASR system.
```
Create another section called ASR with VAD instead of adding to current list. Then the table of content can quickly jump to it. We will add future RNNT too so might as well.

fayejf replied:
Actually, I don't think we need to add another section.  RNNT will go to this tutorial by removing CTC models. And streaming vad asr will go to 

> Streaming / Buffered Automatic Speech Recognition: Streaming_ASR_with_VAD_for_CTC_models

titu1994 reviewed (commented):
Looks better. Minor modifications left.

fayejf commented:
@titu1994 Replied and updated with review comments. Also, make necessary changes for smooth running on colab. Note the new link might be broken due to the 'stable' issue. 
We probably need to revert stable to main for main branch after each release.

titu1994 reviewed (dismissed):
Overall looks good. @tango4j  for final review before merge.
